Mapbox Mobile Events for iOS 1.0.0-beta

alfwatt opened the pull request.

Commit:
Mapbox Mobile Events 1.0-beta1

-> Add MMEVersions.h to store the version info
-> Convert all the #pragma marks
-> I very intentionally MMECommonEventData
-> Remote the MMETimerManager and it's tests, along with MMENamespaceDependencies
-> Only validate the bundle version if it's declared as a semver string

Commit:
Add a target and scheme for building with deprecation warnings enabled
Fix deprecation tag in MMEEventLogger.h

Commit:
Better target name

Commit:
MMEEventLogger → MMELogger
Add tests for MMELogger
Add XCTests to the Deprecation scheme (Cedar tests are Deprecated)
Set log level depending on DEBUG status

Commit:
Check for bundle Ids as well, better filtering and sorting, for more completely consistently reformed user agent strings

rclee commented on MapboxMobileEvents/Categories/NSUserDefaults+MMEConfiguration.m:
```diff
@@ -291,19 +291,27 @@ - (NSString *)mme_userAgentString {
             }
         }
+        
+        // sort the bundleIdentifiers, then use them to build the User-Agent string
```
😙👌

rclee commented on MapboxMobileEvents/MMEMetricsManager.m:
```diff
@@ -232,10 +221,10 @@ - (NSDictionary *)attributes {
+    attributes[MMEEventKeyModel] = MMEEvent.deviceModel;
+    attributes[MMEEventKeyPlatform] = MMEEvent.platformName;
+    attributes[MMEEventKeyOperatingSystem] = MMEEvent.osVersion;
+    attributes[MMEEventKeyDevice] = MMEEvent.deviceModel;
```
Should `deviceModel` be repeated here?

alfwatt replied:
No, no it should not.

/me rustles through diffs

alfwatt commented on MapboxMobileEvents/MMELogger.m:
```diff
@@ -0,0 +1,90 @@
+#if DEBUG
+        self.level = MMELogInfo;
+#else
+        self.level = MMEogFatal;
```
If only there were a release build target I could validate with...

Commit:
Bring back hardwareModel

Commit:
Fix tests, remove Cedar targets, remove Cartfiles
Fix build import cycle in MMEEvent+SystemInfo → NSProcessInfo→SystemInfo

Commit:
Check for and report errors from the session wrapper

Commit:
Make error handling and reporting more consistent for all the network calls

Commit:
Resolved Merge Conflicts, cleanup old schemes, remove MMECategoryLoader

Commit:
Quote mode, unquote

Commit:
`prep` and `lint` are different targerts

dmiluski committed:
[API-Event-Config Testing] Initial Abstraction

Problem:
Current config tests, verify values, but not the differentiation of behaviors based on those config values. Additionally, due to the shared mutable state, Tests must run in a set order with full knowledge/cleanup of the global state around it. This prevents tests from being run in different orders, and covers up an issue in which we can’t run unit tests without knowing global state.

References:
[Why Is Random Order a Big Deal for Test Quality?](https://qualitycoding.org/random-test-order/)

Approach:
Long term, I would like to refactor the architecture here to allow for 1 singleton and instances which can be injected with dependencies rather than outreach to globals.

This initial step introduces a config protocol interface that existing NSUserDefault (real storage conforms to), which can then be injected into individual entities to verify their behavior given config referernce.

Changes:
+ Create MMEEventConfigProviding as a read only config interface
+ Conform NSUserDefaults+MMEConfiguration to new protocol

dmiluski committed:
[DependencyManager] Initial Abstraction

Problem:
Global shared state runs the risk of mutating data out from under you. Migrate dependency manager to an instance wihch can create new location managers.

Although on reading, given it doesn’t look like it’s being configured, I wonder why the need for DependencyManager factory at all?

Changes:
+ Move protocols to groups
- Remove SharedInstance static method
+ Instantiate DependencyManager instance for access

a

dmiluski committed:
[MobileMetricsManager] Initial Abstraction

Goal:
Provide for the ability to inject a MobileMetricsManager allowing for the ability to test individual components on their responsibilities.

Approach:
Given the creation of a metrics manager in MMEEventManager singleton, inject values on init allowing for the eventual test of individual components. This involved creating instances of MMELogger which can be instantiated from MMEEventsManager and passed in with whichever log level configuration may be preferred.

Changes:
+ MMEMetricsManager with dependencies in init
+ MMELocationManager init with dependencies
+ APIClient init with dependencies
+ Add NSBundle Extension to get MME Bundle
+ Configure each at the root MMEEventsManager

dmiluski committed:
[Test Fix] testNSSecureCodingOfMMEEvent

Problem:
Comparison of objects fails

Why:
original vs unarchived failed b/c we are comparing the address of two objects which will be different. For object comparison, Asserts must use EqualObjects for expected comparison.

Change:
Make comparison with XCTAssertEqualObjects
Add XCTAssertNotEqual test

dmiluski committed:
[Invalid Test] testRegionDistanceFilter

Problem:
testRegionDistanceFilter is looking to validate that in the instance of lots of micro location updates, that we’re not overly updating our regions.

Problem:
This test is making comparisons of state synchronously, while the action is asynchronous meaning the comparisons are invalid.

Why:
https://github.com/mapbox/mapbox-events-ios/blob/master/MapboxMobileEvents/MMELocationManager.m#L215

Docs:
 *      This is done asynchronously and may not be immediately reflected in monitoredRegions.
 */
- (void)startMonitoringForRegion:(CLRegion *)region API_AVAILABLE(ios(5.0), macos(10.8)) API_UNAVAILABLE(watchos, tvos);

Approach:
Attempted to inject a spy LocationManager to detect when this was updated, unfortunately this was indeterminant due to LocationManager caching last location. Making this dependent on any other location manager which may have run prior.

Change:
- Remove Test

Alternate Options:
Fully extract out Locationmanager behind a protocol and pass that interface. This would require some implementation behaviors which may not fully match CLLocationManager.

dmiluski committed:
[Bundle] TestBundle Fake Packaging

Goal:
No Need for Host App bundling. We are able to access test bundle with helper category.

Change:
+ Add NSBundle category exposing TestBundle reference
+ Include Test files in this bundle which are accessible regardless of app target allowing for fewer target memberships

Commit:
Contrary assertions

Commit:
Better Assertions

dmiluski committed:
Spelling: ClientID

clietnId -> clientId

dmiluski committed:
[Memory Management] Block Capture List

Goal:
When using blocks, ensure we are not capturing self.

Change:
+ Leverage Weak Self in asynchronous blocks when referenceing self or functions of self. Otherwise, migrate to other categories to support without need for self.

dmiluski committed:
[CI] Enable CI to pass without Carthage Deps

Problem:
CI hosted/run by Bitrise includes Build Steps in a Web GUI. One of which is Carthage bootstrap.

Unfortunately, until the GUI Steps are updated, this will continue to fail.

Approach:
Although we no longer require carthage for installing dependencies the carthage step fails due to lack of Cartfile/Resolved

To allow passing until that time, include empty Cartfile/.resolved

Changes:
+ Add empty Cartfile
+ Don’t ignore .resolved as without it, new builds may reference differerent versions of dependencies.

dmiluski committed:
[Project FIx] Fix Bad Project Rebase

dmiluski committed:
[Tests] Get Tests Running

Goal:
Witht he rebase/cherry pick, some tests weren’t hitting the newly available. Get tests running again.

Changes:
+ Move MMEMetricsManager Tests to instance definitions.
+ ~~Comment out ClientConfigTests until the architecture can be made determinant to verify behavior~~

dmiluski committed:
[Tests] Get Tests Running

Goal:
Witht he rebase/cherry pick, some tests weren’t hitting the newly available. Get tests running again.

Changes:
+ Move MMEMetricsManager Tests to instance definitions.
+ Comment out ClientConfigTests until the architecture can be made determinant to verify behavior

# Conflicts:
#	Tests/MMEAPIConfigTests.m
#	Tests/MMECertPinTests.m

Commit:
We want to know if we're exactly at null-island, not lround there

Commit:
Disorder these tests

dmiluski committed:
URLRequestFactory: Extract Client Responsibility

Goal:
Extract URLRequest Building to a smaller, configurable, testable component.

Approach:
Add URLRequest Factory to share base header/serialization. Allow differentiation via several initializers, but fall through to designated function allowing for re-use/composition of models.

Changes:
+ Add MMENSURLRequestFactory
+ Given configuration via config, Update APIClient initialization
+ Add Mock MMEEventConfig for use in Testing

dmiluski committed:
[Client Decoupling] From EventsManager + Metrics

Problem:
Tight coupling can make it painful if not impossible to test and verify behaviors. In order to allow for clearer undestanding of roles/responsibilities, decouple Client from reaching out to global variable EventManager and Metrics/Logger.

Client Should be responsible for wrapping a URL session with the necessary markup to make service api calls. It should not be resopnsible for dictating tracking behaviors/implementations and logging.

Approach:
1. Remove MetricsManager and EventsManager imports to allow for visiblity into dependencies.
2. Add Blocks which can be configured as needed allowing for decoupling of behaviors.
3. Transition call sites to use block syntax rather than reaching out to singletons.
4. Update Testing to use Block params
5. For Client Testing, inject Block Counter to evaluate that the blocks are called where/as many times as expected.

dmiluski committed:
[Client] Events & Config to Factory/Shared Request

Goal:
1. Provide API calls which fall through a general purpose request function in order to automatically track general purpose network error/bytes.
2. Provide Documented interfaces of each API call
2. Provide Platform Config Abstraction Class in order to factor out decoding/validation logic from Client.
3. Provide the ability to fetch Events Config -> Abstracted model to setup decoupling storage from Client.
4. Document Client Functions

Changes:
- Remove NSURLRequest Category as URLRequests can be made easier with shared support from configured URLRequestFactory
+ Add getConfig/postEvents API Abstraction, transition existing calls to be made through these new layers
+ Add MMEConfig class to support decoding/validation
+ Include Events/Config API URLRequest Tests
- Fix Mock which was referencing a key rather than a url

Commit:
Group the conditional sections in the metrics manager

Commit:
Build drivers for stagecoach

Commit:
use in_port_t for the port number
Revert to CFSocket ref as the type of serverSocket

dmiluski committed:
[EventConfigService] Add Config Polling Controller

Problem:
Currently, the client is responsible for API calls, as well as polling and updating our config.  Although this works, it makes testing difficult as well as may lead to developer confusion when looking for what does what.

Goal:
Extract out this responsibility to a service which can be configured and mocked to verify we’re able to handle any setting sent our way.

Approach:
Add MMEConfigService responsible for the polling

Changes:
+ Add MMEConfigService
+ Integrate service with Client API Call
+ Add test validating polling behavior

Commit:
Unused #defines

Commit:
Whitespace

Commit:
Call delegate when events are enqueued

Commit:
Implement `eventsManager:didSendEvents:`

dmiluski committed:
[URLSession Mocking] via URLProtocol

Goal:
Provide Mechanism to inject configured URLSessionConfiguration with configured protocol classes enabling the easy mocking of responses through our network layer.

This should then provide the ability to test expectations for fetching events config as well as ensuring ConfigService is polling/handling as expected.

Approach:
Leverage URLProtocol for fairly easy mock responses. URLProtocol can be configured as a part of URLSessionConfiguration allowing for opting into handling a request from a local class mock.

References:
https://nshipster.com/nsurlprotocol/
https://developer.apple.com/documentation/foundation/nsurlprotocol

Changes:
+ Providing MMENSURLSessionWrapper initializer to inject dependencies
+ Inject dependencies into client (eg. NSURLSessionConfiguration with protocolClasses)
+ Add EventConfigStubProtocol to host/return stubbed mock on a given url path.
+ Include events-config file (only json)

Note:
Previous file structure did not work as parsing expected only json rather than json + header/response info. Open to alternatives?

By leveraging this approach, we have now ensured that Client/Config Service behave as expects. Other repsonsibilities such as parsing or handling the config values can be validated elsewhere.

dmiluski committed:
[EventReport] Stub API Response

Goal:
Verify client calls/handles response for Event reporting

Changes:
+ Add Stub Protocol
+ Include new test

dmiluski committed:
[EventConfigService] Migrate out of Client

Problem:
Although Client can host this logic, it was adding to much responsibility to the Client class

Goal:
Allow Client to ONLY be responsible for coordinating request/response API calls allowing for testing/configuring of various behaviors more easily through composition.

Change:
+ Remove Config polling responsibility from Client
+ Configure MMEEventsManager to create ConfigService to take on its new roll polling which MMEEventsManager reacts to these changes by updating Defaults upon callbacks.
+ Provide bridge to map/store MMEConfig abstraction class back to NSUserDefault representation.
+ Include MMEConfig Initializer tests

dmiluski committed:
[Memory Management] Fix Weakself Properties

Problem:
When it comes to escaping blocks, we strive to prevent memory cycles which cause leaks. In doing so weakSelf is included in several places in MME SDK. Unfortunately, although we can message nil with function calls, accessing a property of a nil object will cause a crash.

Why:
Memory Access Error

Approach:
- Use stronSelf + checks on blocks
- Use function accessor which plays friendly with accessor due the ability to message nil.

Changes:
- Stub Protocol Completion Block Handler
- Client Completion Block Handler
- Client Document Note on completion block behavior
-

dmiluski committed:
[Build Warnings] Leftover Carthage Cleanup

Problem:
Warning - Linking directory that doesn’t exist

Why:
Given Ceder was removed, we no longer need to reference the search path

Changes:
- Remove sarch path for carthage
- Remove the Ceder dependency leftover in the app

dmiluski committed:
[EventsManager Instantiation] Limit Public Initializers to sharedManager

Problem:
The assumption of init automatically returning a singleton may cause confusion for the most appropriate way to leverage EventsManager leading to many calls to init returning the same thing when a user may expect something different.

Change:
Limit ability to create an EventsManager for internal use.

dmiluski committed:
[MMEPreferences] Abstract Access to NSUserDefaults

Problem:
NSUserDefaults+Category couples both interface and storage. This appraoch has driven the adoption of swapping a static reference to UserDefaults.

Goal:
To allow for clean separation between tests/framework, we need the ability to cleanly separate the interface from the storage so a test does not interfere, or be affected by a host app spinning up it’s own preferences.

Approach: (Class only, not integrated)
Provide Preferences class  as an abstraction around NSUserDefaults that is responsible for providing defaults if values are not set.

Allow for the injection of UserDefaults allowing instances to be created

Changes:
+ Include Public Keys for shared use between NSUserDefaults/Preferences during this transition
+ Create MMEPreferences NSUserDefault Wrapper
+ Build Defaults as part of the accessors so there is no need to register defaults as a two, whitebox knowledge setup

dmiluski committed:
[MMEPreferences] ConfigProviding Support

Goal:
Although NSUserDefaults is fine for storage, exposing models with whitebox knowledge of implemetnation of it’s internals is not idea. Additionally, the requirement to include mme_configuration is also problematic as a consumer may mistakenly forget it. In order to limit these tribal knowledge rules, instad offer a wrapper around user defaults that abstracts away this required knowledge to use.

In an effort to make flows more deterministic without global resets calls scattered through the tests, place Configurations behind an interface allowing for the swapping of Bundle/UseDefault instances for driving flows with injected dependencies.

Expectations:
NSUserDefaults+MMEConfig should still behave as expected for existing SDK consumption.

Approach:
1. Given existence of MMEPreferences, factor out all calls to NSUserDefaults.mme_config from files by injecting read only access to entities that need it via MMEEventConfigProviding.
2. Expose EventConfigProviding for readonly consumption outside of the SDK.
3. Migrate Config Source from NSUserDefaults Category to MMEPreferences wrapper (which allows for instantiated NSUserDefaults sourcing for strong testing)
4. All models should behave the same regardless of interface used, but this sets us up to deprecate the NSUserDefaults interface.

Changes:
+ Conform MMEPreferences to MMEEventConfigProviding
+ Drop mme_ prefixes in ConfigProviding interface as well as adjust consumer references
+ Migrate all consumption of NSUserDefaults to use MMEEventsConfigProviding as a dependency allowing enabling various use case testing rather than reaching out to globals.

dmiluski committed:
[CertificateRevocationList Test] Model/Filtering

Goal:
Enable testing that was disabled (in MMECertPinTests.m)

Approach:
On review, this test wasn’t about logic for CertPinning. Instead it was around logic that was previously hosted in NSUserDefaults+MMEConfiguration, now MMEPreferences.

To enable test, verify the modeling source was accurate as well as the functional calculation once the filter is applied.

Changes:
+ Add MMEPreferencesTest to include Cert/Hash Logical Checks
- Remove Hash logic checks from CertPinTests as it is a model check rather than a pinning logic check
- rename public/china keys to use prefixes to not collide with UserDefaults

dmiluski committed:
[MMEPreferences] Tests

Incorporate tests migrated from NSUserDefaults to MMEPreferenceTests

Approach:
Leverage MMEPrefeernces which can be initialzied with a bundle to pull defaults.

Changes:
+ PreferencesTests

Note:
In some cases, the registered defaults may corrupt the calculated values due to the bundle initially starting with main rather than an injected bundle that MMEPreferences offers.

dmiluski committed:
[Client] Refactor Blocks (OnURLResopnse)

Problem:
Given the stagecoach work, it may be ideal to get access to responses for http response feedback.

Additionally, this revealed some redudancies/code duplication which could be shared in a common path.

Approach:
Reaplce onError/onBytes received with onURLResponse which allows for byte size inspection as well as HTTP errors in a single block. While splitting this out it also seemed useful to separate the reporting of HTTP errors from Serialization/Deserialization errors.

Changes:
+ onLogEvent -> onURLResponse() or onSerializationError() with reporting done outside of client
+ Update Client initializers to fall through sharable initializers leading to less code duplication of defaults.
+ Migrate postMetadata to use designated performRequest which automatically tracks HTTPErrors/Response data sizing via blocks.
+ Update BlockCount + ClientTests to reflect
+ Enable the registration/listening of Serialization/HTTErrors from the MMEEventsManager

dmiluski committed:
[Events Tests] Migrate MMEEvent Tests from Cedar

Goal:
Given Cedar is gone, migrate commented tests to XCTestCases

Changes:
+ Convert Cedar tests to XCTestCases

dmiluski committed:
[Client] Convert Cedar Tests to Async XCTests

Goal:
Migrate Cedar tests to XCTestCases

Given the migration to URLProtocols, sessionWrapFake call validation goes away as it is satisfied by injected URLSession’s mocked response.

Changes:
+ Introduce Error Stub to validate error response block calling
+ Include Asyn testing to validate block call synchronously as well as asynchronously once the call is complete.
- Remove redundant Cedar tests

a

dmiluski committed:
[URLRequestFactory] Migrate some tests from Client

Goal:
Some tests in Client are testing URLFactory Responsibilities. To clearly show what is being tested and where, move request building functions to the factory as well as the tests.

dmiluski committed:
[Remove Cedar] Leftover Project Reference

dmiluski committed:
[MMEMetricsManager] Inject Pending File URL

Problem:
While working on EventManager tests, I ran into a number of failures due to metrics queuing up archived events on each test run. To remedy this side effect between tests, it would be ideal to pass in a pending file path (temporary not/used between tests) so unit tests can execuite without archived metrics affecting tests.

Changes:
+ Add NSURL category to expose easy accessors to test/pending file
+ Add NSURL category to expose accessors to cache/documents.
+ Remove Logger iVar from MetricsManager to decouple logging/events from MMELogger class. Migrate event logging call owner (EventsManager)

Commit:
Can't have a private header in a public header

Commit:
Simplify usage in the README

Commit:
Deprecation → Development
Add private headers to the Development framework

Commit:
Resolve conflicts in rebase from `date/beta+refactor`
